nodeInterface/NodeInterface.go

```diff
@@ -616,57 +616,44 @@ func (n NodeInterface) L2BlockRangeForL1(c ctx, evm mech, l1BlockNum uint64) (ui
 	currentBlockNum := n.backend.CurrentBlock().Number.Uint64()
 	genesis := n.backend.ChainConfig().ArbitrumChainParams.GenesisBlockNum
 
-	type helperStruct struct {
-		low  uint64
-		high uint64
-	}
-
-	searchHelper := func(currentBlock *helperStruct, fetchedMid *helperStruct, target uint64) error {
-		if currentBlock.low < currentBlock.high {
-			// dont fetch midBlockL1Num if its already fetched above
-			mid := arbmath.SaturatingUAdd(currentBlock.low, currentBlock.high) / 2
-			var midBlockL1Num uint64
-			var err error
-			if mid == fetchedMid.low {
-				midBlockL1Num = fetchedMid.high
-			} else {
-				midBlockL1Num, err = n.blockL1Num(mid)
+	storedMids := map[uint64]uint64{}
+	firstL2BlockForL1 := func(target uint64) (uint64, error) {
+		low, high := genesis, currentBlockNum
+		for low < high {
+			mid := arbmath.SaturatingUAdd(low, high) / 2
+			if _, ok := storedMids[mid]; !ok {
+				midBlockL1Num, err := n.BlockL1Num(c, evm, mid)
 				if err != nil {
-					return err
+					return 0, err
 				}
-				fetchedMid.low = mid
-				fetchedMid.high = midBlockL1Num
+				storedMids[mid] = midBlockL1Num
 			}
-			if midBlockL1Num < target {
-				currentBlock.low = mid + 1
+			if storedMids[mid] < target {
+				low = mid + 1
 			} else {
-				currentBlock.high = mid
+				high = mid
 			}
-			return nil
-		}
-		return nil
-	}
-	firstBlock := &helperStruct{low: genesis, high: currentBlockNum}
-	lastBlock := &helperStruct{low: genesis, high: currentBlockNum}
-	// in storedMid low corresponds to value mid and high corresponds to midBlockL1Num inside searchHelper
-	storedMid := &helperStruct{low: currentBlockNum + 1}
-	var err error
-	for firstBlock.low < firstBlock.high || lastBlock.low < lastBlock.high {
-		if err = searchHelper(firstBlock, storedMid, l1BlockNum); err != nil {
-			return 0, 0, err
-		}
-		if err = searchHelper(lastBlock, storedMid, l1BlockNum+1); err != nil {
-			return 0, 0, err
 		}
+		return high, nil
 	}
-	if err := n.matchL2BlockNumWithL1(firstBlock.high, l1BlockNum); err != nil {
+
+	firstBlock, err := firstL2BlockForL1(l1BlockNum)
+	if err != nil {
+		return 0, 0, err
+	}
+	lastBlock, err := firstL2BlockForL1(l1BlockNum + 1)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if err := n.matchL2BlockNumWithL1(c, evm, firstBlock, l1BlockNum); err != nil {
 		return 0, 0, err
 	}
-	if err := n.matchL2BlockNumWithL1(lastBlock.high, l1BlockNum); err != nil {
-		lastBlock.high -= 1
-		if err = n.matchL2BlockNumWithL1(lastBlock.high, l1BlockNum); err != nil {
+	if err := n.matchL2BlockNumWithL1(c, evm, lastBlock, l1BlockNum); err != nil {
```

Comment:
if you run the binary search you would realize that just matching lastblock-1 would return wrong values when the query is the latest L1 block number in which case the L2 looks something like this and we are querying for l1Num=x `[x-2 , x-2 , x-2 , x-1 , x-1 , x , x , x , x]`, correct answer here would be lastblock and not lastblock-1.

Reply:
That is a good point, so basically the binary search above doesn't really do what one would expect it to do: 
Your binary search for some target `t` should be returning minimum l2 block index `l2BlockIdx` such that `L1IndexOf(l2BlockIdx) >= t`, but it's actually not doing that.

With your example, where `l1Num=x` second binary search run: `lastBlock, err := firstL2BlockForL1(l1BlockNum + 1)`
will return block that is included in `l1BlockNum` not in `l1BlockNum+1` (because that block doesn't exist yet).

Please do that edge case logic inside binary search (that is what I meant in https://github.com/OffchainLabs/nitro/pull/1860#discussion_r1330445050 by "check ends"): before beginning the loop, check, `if n.blockL1Num(high) < target` then return high +1.

I think that would be clear way because you contain corner cases in the binary search and won't have to handle it outside.

Reply:
It makes sense to handle it inside search loop too, but when I wrote the implementation I was trying to minimize the queries (this does add one extra check in average case). Moving the check inside binary search does have a benefit too i.e we skip one entire loop when the user queries for `l1BlockNum=latest`. Thanks for suggesting the change